File: DesignPatterns/01_GOF23/DP_examples/Go/02_Structural/Bridge/implementor.go
package Bridge

import "fmt"

// 颜色枚举
type Color int

const (
	Unknown Color = iota
	Red
	Green
	Blue
)

func (c Color) String() string {
	switch c {
	case Red:
		return "red"
	case Green:
		return "green"
	case Blue:
		return "blue"
	default:
		return "unknown"
	}
}

// 实现接口
type DrawAPI interface {
	Draw()
	Fill(color Color)
}

type drawAPI_Impl struct{}

func (d *drawAPI_Impl) Draw() {}
func (d *drawAPI_Impl) Fill(color Color) {
	fmt.Println("Fill with", color)
}

// 圆形实现
type Circle struct {
	drawAPI_Impl
}

func (c *Circle) Draw() {
	fmt.Println("Draw a Circle")
}

// 正方形实现
type Square struct {
	drawAPI_Impl
}

func (s *Square) Draw() {
	fmt.Println("Draw a Square")
}

// 矩形实现
type Rectangle struct{}

func (r *Rectangle) Draw() {
	fmt.Println("Draw a Rectangle")
}

// 大圆形实现
type BigCircle struct {
	Circle
}

func (bc *BigCircle) Draw() {
	fmt.Println("Draw a Big Circle")
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/02_Structural/Bridge/Implementor.lua
---@enum Color
Color = {
    unknown = 0,
    red = 1,
    green = 2,
    blue = 3
}
---@param color Color
---@return string
local function ColorToString(color)
    if color == Color.red then
        return "red"
    elseif color == Color.green then
        return "green"
    elseif color == Color.blue then
        return "blue"
    else
        return "unknown"
    end
end

---@class DrawAPI
DrawAPI = {}
function DrawAPI:Draw()
    error("Subclass must implement draw method")
end
---@param color Color
function DrawAPI:Fill(color)
    print("Fill with " .. ColorToString(color))
end

-- 创建 DrawAPI 实例
---@return DrawAPI
function DrawAPI:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

---@class Circle : DrawAPI
Circle = DrawAPI:new()
function Circle:Draw()
    print("Draw a Circle")
end

---@class Square : DrawAPI
Square = DrawAPI:new()
function Square:Draw()
    print("Draw a Square")
end

---@class Rectangle : DrawAPI
Rectangle = DrawAPI:new()
function Rectangle:Draw()
    print("Draw a Rectangle")
end

---@class BigCircle : Circle
BigCircle = Circle:new()
function BigCircle:Draw()
    print("Draw a Big Circle")
end
